### PR TITLE
[pkg/telemetryquerylanguage] Add enums to TQL

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
This PR adds Enums to the TQL.  This is necessary before the transform processor and be updated to use the package.

~The changes are copied from the current Enum implementation in the transformprocessor.~

This implementation improves upon the transform processor's version by separating Enums from Paths.  This allows us to validate Enums and returns errors if there is no matching Symbol.

**Link to tracking Issue:** 
#11751

**Testing:** 
Updated unit tests

**Documentation:** 
Updated README.

### CODE REVIEW

File: pkg/telemetryquerylanguage/tql/functions.go
Comment (kentquirk) [inline pkg/telemetryquerylanguage/tql/functions.go:24]: What do you think about making EnumParser returning an error rather than just a bool? That gives a place for the implemented EnumParser to be more informative than just "NO". ("'IFNO' not found -- did you mean 'INFO'?" is one example).

File: pkg/telemetryquerylanguage/tql/functions_test.go
Comment (kentquirk) [inline pkg/telemetryquerylanguage/tql/functions_test.go:341]: I think we need the test for what happens if this enum isn't found.

File: pkg/telemetryquerylanguage/tql/condition_test.go
Comment (kentquirk) [inline pkg/telemetryquerylanguage/tql/condition_test.go:98]: We could use a couple more tests in this context:
* Enum on the right
* Enum not found

**Review by kentquirk — COMMENTED** [review]
I think this is generally a good and useful change. I'd like to see the extra tests and to consider whether a more robust error scheme would be handy.

(And selfishly, I'd like to see this rebased after the merge of the and-or branch.)

File: pkg/telemetryquerylanguage/tql/condition_test.go
Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/condition_test.go:98]: Enum not found is an interesting scenario. I'm away from my computer right now but if I remember my implementation correctly if the Path doesn't match any Enums it'll then be parsed as a Path, which will fail. This would be caught during config validation. 

I can add a test scenario to verify this.

File: pkg/telemetryquerylanguage/tql/parser_test.go
Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/parser_test.go:412]: I can add more interesting values and tests.

File: pkg/telemetryquerylanguage/tql/functions.go
Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: The tricky part would be trying to differentiate between an enum that didn't match in the symbol table and a Path that needs to be interpreted as part as a Path.

This could be solved if we can make the grammar smart enough to differentiate between a Path and an Enum (as right now they are the same according to the lexer). I started out with that but ended up with this solution bc I didn't like having to add something before, after, or around the Enum to differentiate it from a Path.

**Comment by TylerHelmuth** [comment]
@kentquirk I added more tests but have not modified the Enum API.  Since Enums are being treated as fancy paths, the ParseEnum function doesn't know if the provided Ident is intended to be an Enum or not; it only know that it does or does not match something in the symbol table.  If the provided path does not match anything from `ParseEnum` or `PathExpressionParser` then `ParseQueries` will return an error.  

I'm open to ideas on how to separate Enum and Path in the grammar, but I don't love the idea of having to add anything to the symbol.  Being able to interact with Enums like `... where kind == SPAN_KIND_INTERNAL` is very satisfying.

File: pkg/telemetryquerylanguage/tql/functions.go
Comment (kentquirk) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: I probably wouldn't say that it has to have a marker for Enum. Right now, a Path is anything that contains a dot `.`, and an Invocation has to contain parens. So we could, for example, constrain Enum to *not* contain a dot. Personally, I'd prefer saying that it has to be something that looks like an Ident (`[A-Za-z_][A-Za-z0-9_]`), and maybe even specifically an uppercase identifier (`[A-Z_][A-Z0-9_]`). 

What do you think?

Comment (kentquirk) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: OK, I read the definition too quickly -- but does it have to be that way? There are two questions here:
- Could we constrain Enum (say) to be an Ident? I'd like to, even if Paths have to be the way they are.
- Could we constrain Paths to be more restrictive? I am probably missing something, but all of the examples I know of contain a `.` -- the grammar could use a `+` instead of `*` and enforce this restriction.

I'm sorry if it feels like I'm being nitpicky here -- I am thinking that in a general-purpose language, having more restrictions on symbols helps in writing grammars that can actually be helpful in their error messages, and if we continue with a "Path is anything" model, then it will likely complicate things later. These sorts of restrictions are easy to remove, but difficult to add.

Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: I don't think the path can be more restrictive, thats the issue I ran into when I force implemented this in the transform processor 😞   

Path is the connection between the grammar and the signal's context.  For the transformprocessor it is how we represent the underlying signal's pdata model. So when we want to access a span's name we use the path `name`.  When we want to access a data point's start time you use the path `start_time_unix_nano`. 

There are other aspects of the data model we need to access such as `resource`, `instrumentation_scope` and `metric`. We use the dots to represent these hierarchical relationships, but they are the exception, not the norm.  Most pdata fields are accessed via a path without a dot.

Looking through `traces.go`, `metrics.go`, and `logs.go` I think we **do** represent all pdata paths as lowercase.  We might be able to differentiate between an Enum and a Path if we can enforce all Enums to be UPPERCASE and all paths to be lowercase.  I will look into that.

Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: At the moment both Path and Invocation rely on the @Ident token which allows both cases.  I think to enforce Paths being lower case and Enums being upper case we'd need new tokens to enforce that, but they would conflict with @Ident right?  Anything that would match an only UPPERCASE token or only lowercase token would also match the @Ident token.

Comment (kentquirk) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: Yes, Ident would have to change. I would like to impose these restrictions, but I won't insist on it.

Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: Any ideas on how to change Ident?

Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: We might be able to take advantage of this capability in the grammar library:

`"...":<identifier> Match the literal, specifying the exact lexer token type to match.`

Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/functions.go:24]: NVM, it would have to be a literal, which isn't applicable here

Comment (TylerHelmuth) [reply pkg/telemetryquerylanguage/tql/functions.go]: @kentquirk I think I got it.  I removed Ident and made uppercase and lowercase tokens.  Now the grammar looks like:

- Uppercase for Enum
- Lowercase for Path
- (Uppercase | Lowercase)+ for Invocation.Function

This allows EnumParser to return an error like PathExpressionParser.  All other aspects of the grammar should rename the same.  Path are forced to be lowercase now, but I don't think that is a problem since the transform processor has already established the patter of Paths being lower_snake_case.

**Review by kentquirk — APPROVED** [review]
This makes me happy now. I have no further comments.

**Comment by TylerHelmuth** [comment]
@open-telemetry/collector-contrib-approvers please review.  I'd love to get this PR in so I can complete https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12221, all hopefully before Wednesday's release.  There are some users who are interested in the new AND and OR where clause capabilities that are in the library but not in the processor.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**
